src/modules/kvs/kvs_checkpoint.c

```diff
@@ -0,0 +1,230 @@
+    flux_t *h;
+    struct kvsroot *root_primary;
+    double checkpoint_period;   /* in seconds */
+    flux_watcher_t *checkpoint_w;
+    kvs_checkpoint_txn_cb txn_cb;
+    void *txn_cb_arg;
+    int last_checkpoint_seq;
+};
+
+static int checkpoint_period_parse (const flux_conf_t *conf,
+                                    flux_error_t *errp,
+                                    double *checkpoint_period)
+{
+    flux_error_t error;
+    const char *str = NULL;
+
+    if (flux_conf_unpack (conf,
+                          &error,
+                          "{s?{s?s}}",
+                          "kvs",
+                          "checkpoint-period", &str) < 0) {
+        errprintf (errp,
+                   "error reading config for kvs: %s",
+                   error.text);
+        return -1;
+    }
+
+    if (str) {
+        if (fsd_parse_duration (str, checkpoint_period) < 0) {
+            errprintf (errp,
+                       "invalid checkpoint-period config: %s",
+                       str);
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+int kvs_checkpoint_config_parse (kvs_checkpoint_t *kcp,
+                                 const flux_conf_t *conf,
+                                 flux_error_t *errp)
+{
+    if (kcp) {
+        double checkpoint_period = kcp->checkpoint_period;
+        if (checkpoint_period_parse (conf, errp, &checkpoint_period) < 0)
+            return -1;
+        kcp->checkpoint_period = checkpoint_period;
+    }
+    return 0;
+}
+
+int kvs_checkpoint_reload (kvs_checkpoint_t *kcp,
+                           const flux_conf_t *conf,
+                           flux_error_t *errp)
+{
+    if (kcp) {
+        double checkpoint_period = kcp->checkpoint_period;
+        if (checkpoint_period_parse (conf,
+                                     errp,
+                                     &checkpoint_period) < 0)
+            return -1;
+
+        if (checkpoint_period != kcp->checkpoint_period) {
+            kcp->checkpoint_period = checkpoint_period;
+            flux_watcher_stop (kcp->checkpoint_w);
+
+            if (kcp->root_primary
+                && kcp->checkpoint_period > 0.0) {
+                flux_timer_watcher_reset (kcp->checkpoint_w,
+                                          kcp->checkpoint_period,
+                                          kcp->checkpoint_period);
+                flux_watcher_start (kcp->checkpoint_w);
+            }
+        }
+    }
+    return 0;
+}
+
+
+static void checkpoint_cb (flux_reactor_t *r,
+                           flux_watcher_t *w,
+                           int revents,
+                           void *arg)
+{
+    kvs_checkpoint_t *kcp = arg;
+    char *name = NULL;
+    json_t *ops = NULL;
+
+    /* if no changes to root since last checkpoint-period, do
+     * nothing */
+    if (kcp->last_checkpoint_seq == kcp->root_primary->seq)
+        return;
+
+    if (asprintf (&name, "checkpoint-period.%u", kcp->root_primary->seq) < 0
```

Comment:
It would probably be better to use a static buffer instead of `asprintf()` here since the string is of predictable length.